source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp

```diff
@@ -0,0 +1,318 @@
+                            status = shader.setParameter(global.c_str(), mipCount);
+                        }
+                    }
+                }
+            }
+
+            // Environment matrix
+            else if (global == ENVIRONMENT_MATRIX_PARAMETER)
+            {
+                if (parameterList.indexOf(global.c_str()) >= 0)
+                {
+                    const float yRotationPI[4][4]{
+                        -1, 0, 0, 0,
+                        0, 1, 0, 0,
+                        0, 0, -1, 0,
+                        0, 0, 0, 1
+                    };
+                    MFloatMatrix matrix(yRotationPI);
+                    status = shader.setParameter(global.c_str(), matrix);
+                }
+            }
+        }
+    }
+} // anonymous namespace
+
+template <class BASE>
+MaterialXShadingNodeImpl<BASE>::MaterialXShadingNodeImpl(const MObject& obj)
+	: BASE(obj)
+	, _object(obj)
+{
+}
+
+template <class BASE>
+MaterialXShadingNodeImpl<BASE>::~MaterialXShadingNodeImpl()
+{
+	// TODO: Free sampler state here!
+}
+
+template <class BASE>
+MString
+MaterialXShadingNodeImpl<BASE>::fragmentName() const
+{
+	MStatus status;
+	MFnDependencyNode depNode(_object, &status);
+	const auto* const node = dynamic_cast<MaterialXNode*>(depNode.userNode());
+    const MaterialXData* const data = node ? node->materialXData.get() : nullptr;
+    return data ? data->getFragmentName().c_str() : "";
+}
+
+template <class BASE>
+void MaterialXShadingNodeImpl<BASE>::updateDG()
+{
+}
+
+template <class BASE>
+void
+MaterialXShadingNodeImpl<BASE>::updateShader(
+    MHWRender::MShaderInstance& shader,
+    const MHWRender::MAttributeParameterMappingList& mappings
+)
+{
+    MStatus status;
+    MFnDependencyNode depNode(_object, &status);
+    const auto* const node = dynamic_cast<MaterialXNode*>(depNode.userNode());
+    if (!node)
+    {
+        return;
+    }
+
+    // Get the parameter list fo checking against.
+    MStringArray parameterList;    
+    shader.parameterList(parameterList);
+
+    // Set up image file name search path. Assume we are using built in images located in resource path
+    // TODO: Be able to add more image search paths.
+    static std::string IMAGE_FOLDER("Images");
+    MaterialX::FileSearchPath imageSearchPath(Plugin::instance().getResourcePath() / MaterialX::FilePath(IMAGE_FOLDER));
+
+    const MaterialX::StringVec& globals = node->materialXData->getGlobalsList();
+
+    // Bind environment lighting
+    // TODO: These should be options
+    std::string envRadiancePath = "san_giuseppe_bridge.hdr";
+    std::string envIrradiancePath = "san_giuseppe_bridge_diffuse.hdr";
+    ::bindEnvironmentLighting(shader, globals, parameterList, imageSearchPath, 
+                            envRadiancePath, envIrradiancePath);
+
+    MaterialX::DocumentPtr document = node->materialXData->getDocument();
+	const MaterialX::StringMap& inputs = node->materialXData->getPathInputMap();
```

Comment:
Looks like all your tabbing is still messed up. Should be 4 spaces, no hard-tabs.

Reply:
This file was copied from MaterialXTextureOverride, the tabs are left over from there. I'll clean up.